joomla/joomla-cms
# [4.3] Fix javascript error on frontend MFA captive login page

Pull Request for Issue #40273 .

### Summary of Changes

This pull request (PR) moves the `.querySelector('button')` in the "two-factor-focus.es6.js" script to the right place to fix a javascript error on the MFA captive login page for the site (frontend).

The remaining code in that function is already so that it does not need the removed part.

### Testing Instructions

Apply the changes from this PR and then run `npm ci` or `npm run build:js` when on a development environment.

When on a regular installation where you have no npm, use the update package or the custom update URL created by drone for this PR.

After having applied the changes, clear your browser cache to make sure you are not testing the old, unmodified javascript.

1. Create a user account with MFA active who can login to backend and frontend.
The user should have activated at least one MFA method which requires to enter a numeric verification code, e.g. TOTP, and another one which doesn't require such a code (Webauthn e.g with Windows Hello).
2. Login to the frontend with this account with username and password.
3. Check if depending on the MFA method, the input field for the verification code has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are js errors.
Result: See section "Actual result BEFORE applying this Pull Request" without the PR, and "Expected result AFTER applying this Pull Request" with the PR.
4. Enter a verification code if necessary and use the validation button to finally log in.
5. Login to the backend with this account with username and password.
6. Check if depending on the MFA method, the input field for the verification code or the button to validate with the authenticator has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are js errors.
Result: The MFA captive login page for the backend works with this PR on the same way as without this PR, i.e. nothing is broken by this PR, no javascript errors related to the "two-factor-focus" in both cases.
7. Enter a verification code if necessary and use the validation button to finally log in.

### Actual result BEFORE applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, no button does have focus because the javascript modified by this PR here is not loaded, see https://github.com/joomla/joomla-cms/blob/4.3-dev/components/com_users/tmpl/captive/default.php#L25-L28 , i.e. this works as expected.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, you get a Javascript error "Uncaught TypeError: document.getElementById(...) is null" in the browser console, and no input field has focus.
When "Debug System" is on in Global Configuration so the unminified javascript is loaded, you will see the error happens here: https://github.com/joomla/joomla-cms/blob/4.3-dev/build/media_source/com_users/js/two-factor-focus.es6.js#L12

On the MFA captive login page for the backend everything works as expected:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, the button to validate with the authenticator has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus.

### Expected result AFTER applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, everything works as  before applying this pull request, i.e. no button has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus. There are no javascript error in browser console related to the "two-factor-focus" script.

On the MFA captive login page for the backend: Everything works as well as before applying this pull request.

### Link to documentations
Please select:
